openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMBinaryClassificationModelTrainer.java

```diff
@@ -58,43 +54,66 @@
+     *     <li>When the chunk is full, ChunkedArray adds another chunk and inserts the value there.</li>
+     *     <li>Repeat from 2 until the input is exhausted.</li>
+     * </ol>
+     * <p>
+     * Computing the dataset size from the ChunkedArray is an O(1) operation.
+     * As all chunks have the same size:
+     * size = (num_chunks-1)*chunk_size + num_elements_in_last_chunk
      *
-     * @param dataset Train dataset
-     * @param params LightGBM model parameters
+     * @param dataset             Train dataset.
+     * @param instancesPerChunk   Number of instances for each train chunk in C++.
+     * @param params              LightGBM model parameters.
      * @param outputModelFilePath Output filepath for the model file in .txt format.
      */
-    static void fit(final Dataset dataset, final Map<String, String> params, final Path outputModelFilePath) {
+    static void fit(final Dataset dataset,
+                    final long instancesPerChunk,
+                    final Map<String, String> params,
+                    final Path outputModelFilePath) {
 
         final DatasetSchema schema = dataset.getSchema();
         final int numFeatures = schema.getPredictiveFields().size();
-        // This shouldn't be needed, but at this level of abstraction, Dataset does not provide a method for size:
-        final int numInstances = Iterators.size(dataset.getInstances());
-        logger.debug("The train dataset has {} instances.", numInstances);
-        final List<Integer> categoricalFeatureIndicesWithoutLabel = getCategoricalFeaturesIndicesWithoutLabel(schema);
-        final String trainParams = getLightGBMTrainParamsString(params, categoricalFeatureIndicesWithoutLabel);
+        final String trainParams = getLightGBMTrainParamsString(params,
+                getCategoricalFeaturesIndicesWithoutLabel(schema));
         final int numIterations = parseInt(params.get(NUM_ITERATIONS_PARAMETER_NAME));
         logger.debug("LightGBM model trainParams: {}", trainParams);
 
-        final SWIGTrainResources swigTrainResources = new SWIGTrainResources(numInstances, numFeatures);
+        final SWIGTrainData swigTrainData = new SWIGTrainData(
+                numFeatures,
+                instancesPerChunk);
+        final SWIGTrainBooster swigTrainBooster = new SWIGTrainBooster();
 
         /// Create LightGBM dataset
-        createTrainDataset(dataset, numFeatures, numInstances, trainParams, swigTrainResources);
+        createTrainDataset(dataset, numFeatures, trainParams, swigTrainData);
 
         /// Create Booster from dataset
-        createBoosterStructure(swigTrainResources, trainParams);
-        trainBooster(swigTrainResources.swigBoosterHandle, numIterations);
+        createBoosterStructure(swigTrainBooster, swigTrainData, trainParams);
+        trainBooster(swigTrainBooster.swigBoosterHandle, numIterations);
```

Comment:
Ideally, as now Data and Booster are split, we'd move these operations to the specific SWIGTrainData and SWIGTrainBooster classes in a separate PR.